File: android/src/main/java/com/getcapacitor/community/database/sqlite/SQLite/UtilsDelete.java
package com.getcapacitor.community.database.sqlite.SQLite;

import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLStatement.addPrefixToWhereClause;
import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLStatement.extractForeignKeyInfo;
import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLStatement.flattenMultilineString;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.json.JSONException;
import org.json.JSONObject;

public class UtilsDelete {

    private final UtilsSQLStatement _statUtil = new UtilsSQLStatement();

    public static class ReferenceResult {

        String tableWithRefs;
        List<String> retRefs;
    }

    public static class ForeignKeyInfo {

        String tableName;
        List<String> forKeys;
        List<String> refKeys;
        String action;

        public ForeignKeyInfo() {
            this.tableName = "";
            this.forKeys = new ArrayList<>();
            this.refKeys = new ArrayList<>();
            this.action = "NO ACTION";
        }

        public ForeignKeyInfo(List<String> forKeys, String tableName, List<String> refKeys, String action) {
            this.tableName = tableName;
            this.forKeys = forKeys;
            this.refKeys = refKeys;
            this.action = action;
        }

        public String getTableName() {
            return tableName;
        }

        public String getAction() {
            return action;
        }

        public List<String> getForKeys() {
            return forKeys;
        }

        public List<String> getRefKeys() {
            return refKeys;
        }
    }

    public static class UpdateResults {

        private String setStmt;
        private String updWhereStmt;

        public UpdateResults() {
            this.setStmt = "";
            this.updWhereStmt = "";
        }

        public UpdateResults(String setStmt, String updWhereStmt) {
            this.setStmt = setStmt;
            this.updWhereStmt = updWhereStmt;
        }

        public String getSetStmt() {
            return setStmt;
        }

        public String getUpdWhereStmt() {
            return updWhereStmt;
        }
    }

    public static boolean findReferencesAndUpdate(
        Database mDB,
        String tableName,
        String whereStmt,
        String[] initColNames,
        ArrayList<Object> values
    ) throws Exception {
        try {
            boolean retBool = true;
            ReferenceResult result = getReferences(mDB, tableName);
            List<String> references = result.retRefs;
            String tableNameWithRefs = result.tableWithRefs;

            if (references.size() <= 0) {
                return retBool;
            }

            if (tableName.equals(tableNameWithRefs)) {
                return retBool;
            }

            for (String ref : references) {
                ForeignKeyInfo foreignKeyInfo = extractForeignKeyInfo(ref);

                String refTable = foreignKeyInfo.tableName;
                if (refTable.isEmpty() || !refTable.equals(tableName)) {
                    continue;
                }

                List<String> withRefsNames = foreignKeyInfo.forKeys;
                List<String> colNames = foreignKeyInfo.refKeys;

                if (colNames.size() != withRefsNames.size()) {
                    throw new Error("findReferencesAndUpdate: mismatch length");
                }

                String action = foreignKeyInfo.action;
                if (action.equals("NO_ACTION")) {
                    continue;
                }

                String updTableName = tableNameWithRefs;
                List<String> updColNames = withRefsNames;

                UpdateResults results = new UpdateResults();

                if (!checkValuesMatch(withRefsNames.toArray(new String[0]), initColNames)) {
                    Map<String, Object> relatedItemsResult = searchForRelatedItems(
                        mDB,
                        updTableName,
                        tableName,
                        whereStmt,
                        withRefsNames.toArray(new String[0]),
                        colNames.toArray(new String[0]),
                        values
                    );

                    if (
                        ((List<String>) relatedItemsResult.get("relatedItems")).size() == 0 &&
                        ((String) relatedItemsResult.get("key")).length() == 0
                    ) {
                        continue;
                    }

                    if (!updTableName.equals(tableName)) {
                        switch (action) {
                            case "RESTRICT":
                                results = upDateWhereForRestrict(relatedItemsResult);
                                break;
                            case "CASCADE":
                                results = upDateWhereForCascade(relatedItemsResult);
                                break;
                            default:
                                results = upDateWhereForDefault(withRefsNames, relatedItemsResult);
                                break;
                        }
                    }
                } else {
                    throw new Error("Not implemented. Please transfer your example to the maintainer");
                }

                if (!results.getSetStmt().isEmpty() && !results.getUpdWhereStmt().isEmpty()) {
                    executeUpdateForDelete(mDB, updTableName, results.getUpdWhereStmt(), results.getSetStmt(), updColNames, values);
                }
            }
            return retBool;
        } catch (Exception error) {
            String msg = error.getMessage() != null ? error.getMessage() : error.toString();
            throw new Exception(msg);
        }
    }

    public static ReferenceResult getReferences(Database mDB, String tableName) throws Exception {
        String sqlStmt =
            "SELECT sql FROM sqlite_master " +
            "WHERE sql LIKE('%FOREIGN KEY%') AND sql LIKE('%REFERENCES%') AND " +
            "sql LIKE('%" +
            tableName +
            "%') AND sql LIKE('%ON DELETE%');";

        try {
            JSArray references = mDB.selectSQL(sqlStmt, new ArrayList<>());
            ReferenceResult referenceResult = new ReferenceResult();
            List<String> retRefs = new ArrayList<>();
            String tableWithRefs = "";

            if (references.length() > 0) {
                Map<String, Object> result = getRefs(references.getJSONObject(0).getString("sql"));
                retRefs = (List<String>) result.get("foreignKeys");
                tableWithRefs = (String) result.get("tableName");
            }

            referenceResult.tableWithRefs = tableWithRefs;
            referenceResult.retRefs = retRefs;

            return referenceResult;
        } catch (Exception e) {
            String error = e.getMessage() != null ? e.getMessage() : e.toString();
            String msg = "getReferences: " + error;
            throw new Exception(msg);
        }
    }

    public static Map<String, Object> getRefs(String sqlStatement) throws Exception {
        Map<String, Object> result = new HashMap<>();
        String tableName = "";
        List<String> foreignKeys = new ArrayList<>();
        String statement = flattenMultilineString(sqlStatement);

        try {
            // Regular expression pattern to match the table name
            String tableNamePattern = "CREATE\\s+TABLE\\s+(\\w+)\\s+\\(";
            Pattern tableNameRegex = Pattern.compile(tableNamePattern);
            Matcher tableNameMatcher = tableNameRegex.matcher(statement);
            if (tableNameMatcher.find()) {
                tableName = tableNameMatcher.group(1);
            }

            // Regular expression pattern to match the FOREIGN KEY constraints
            String foreignKeyPattern =
                "FOREIGN\\s+KEY\\s+\\([^)]+\\)\\s+REFERENCES\\s+(\\w+)\\s*\\([^)]+\\)\\s+ON\\s+DELETE\\s+(CASCADE|RESTRICT|SET\\s+DEFAULT|SET\\s+NULL|NO\\s+ACTION)";
            Pattern foreignKeyRegex = Pattern.compile(foreignKeyPattern);
            Matcher foreignKeyMatcher = foreignKeyRegex.matcher(statement);
            while (foreignKeyMatcher.find()) {
                String foreignKey = foreignKeyMatcher.group(0);
                foreignKeys.add(foreignKey);
            }
        } catch (Exception e) {
            String msg = "getRefs: Error creating regular expression: " + e.toString();
            throw new Exception(msg);
        }

        result.put("tableName", tableName);
        result.put("foreignKeys", foreignKeys);
        return result;
    }

    public static boolean checkValuesMatch(String[] array1, String[] array2) {
        for (String value : array1) {
            boolean found = false;
            for (String item : array2) {
                if (value.equals(item)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                return false;
            }
        }
        return true;
    }

    public static Map<String, Object> searchForRelatedItems(
        Database mDB,
        String updTableName,
        String tableName,
        String whStmt,
        String[] withRefsNames,
        String[] colNames,
        ArrayList<Object> values
    ) throws Exception {
        List<Object> relatedItems = new ArrayList<>();
        String key = "";
        String[] t1Names = new String[withRefsNames.length];
        String[] t2Names = new String[colNames.length];

        for (int i = 0; i < withRefsNames.length; i++) {
            t1Names[i] = "t1." + withRefsNames[i];
            t2Names[i] = "t2." + colNames[i];
        }

        try {
            // addPrefix to the whereClause and swap colNames with  withRefsNames
            String whereClause = addPrefixToWhereClause(whStmt, colNames, withRefsNames, "t2.");
            // look at the whereclause and change colNames with  withRefsNames
            if (whereClause.endsWith(";")) {
                whereClause = whereClause.substring(0, whereClause.length() - 1);
            }

            StringBuilder resultString = new StringBuilder();
            for (int index = 0; index < t1Names.length; index++) {
                resultString.append(t1Names[index]).append(" = ").append(t2Names[index]);
                if (index < t1Names.length - 1) {
                    resultString.append(" AND ");
                }
            }

            String sql =
                "SELECT t1.rowid FROM " +
                updTableName +
                " t1 " +
                "JOIN " +
                tableName +
                " t2 ON " +
                resultString.toString() +
                " " +
                "WHERE " +
                whereClause +
                " AND t1.sql_deleted = 0;";

            JSArray jsVals = mDB.selectSQL(sql, values);
            if (jsVals.length() > 0) {
                List<Map<String, Object>> mVals = JSArrayToJavaListMap(jsVals);
                key = mVals.get(0).keySet().iterator().next();
                relatedItems.addAll(mVals);
            }
            Map<String, Object> result = new HashMap<>();
            result.put("key", key);
            result.put("relatedItems", relatedItems);
            return result;
        } catch (Exception error) {
            String msg = error.getMessage() != null ? error.getMessage() : error.toString();
            throw new Exception(msg);
        }
    }

    public static List<Map<String, Object>> JSArrayToJavaListMap(JSArray jsArray) throws JSONException {
        List<Map<String, Object>> listMap = new ArrayList<>();

        for (int i = 0; i < jsArray.length(); i++) {
            JSObject jsObject = (JSObject) jsArray.get(i); // Assuming each element is an object
            Map<String, Object> map = new HashMap<>();

            // Extract key-value pairs from the JSObject and put them into the Map
            for (Iterator<String> it = jsObject.keys(); it.hasNext();) {
                String key = it.next();
                map.put(key, jsObject.get(key)); // Convert JSValue to Java object
            }

            listMap.add(map); // Add the map to the list
        }

        return listMap; // Return the List<Map<String, Object>>
    }

    public static UpdateResults upDateWhereForRestrict(Map<String, Object> results) throws Exception {
        try {
            if (((List<?>) results.get("relatedItems")).size() > 0) {
                String msg = "Restrict mode related items exist, please delete them first";
                throw new Exception(msg);
            }
            return new UpdateResults();
        } catch (Exception error) {
            String msg = error.getMessage() != null ? error.getMessage() : "";
            throw new Exception(msg);
        }
    }

    public static UpdateResults upDateWhereForCascade(Map<String, Object> results) throws Exception {
        String setStmt = "";
        String uWhereStmt = "";

        try {
            String key = (String) results.get("key");
            List<Object> cols = new ArrayList<>();
            List<Map<String, Object>> relatedItems = (List<Map<String, Object>>) results.get("relatedItems");

            for (Map<String, Object> relItem : relatedItems) {
                Object mVal = relItem.get(key);
                if (mVal != null) {
                    cols.add(mVal);
                }
            }

            setStmt += "sql_deleted = 1";

            // Create the where statement
            StringBuilder uWhereStmtBuilder = new StringBuilder("WHERE " + key + " IN (");
            for (Object col : cols) {
                uWhereStmtBuilder.append(col).append(",");
            }
            if (uWhereStmtBuilder.toString().endsWith(",")) {
                uWhereStmtBuilder.deleteCharAt(uWhereStmtBuilder.length() - 1);
            }
            uWhereStmtBuilder.append(");");
            uWhereStmt = uWhereStmtBuilder.toString();
        } catch (Exception error) {
            String msg = error.getMessage() != null ? error.getMessage() : "";
            throw new Exception(msg);
        }
        return new UpdateResults(setStmt, uWhereStmt);
    }

    public static UpdateResults upDateWhereForDefault(List<String> withRefsNames, Map<String, Object> results) throws Exception {
        String setStmt = "";
        String uWhereStmt = "";

        try {
            String key = (String) results.get("key");
            List<Object> cols = new ArrayList<>();
            List<Map<String, Object>> relatedItems = (List<Map<String, Object>>) results.get("relatedItems");

            for (Map<String, Object> relItem : relatedItems) {
                Object mVal = relItem.get(key);
                if (mVal != null) {
                    cols.add(mVal);
                }
            }

            // Create the set statement
            for (String name : withRefsNames) {
                setStmt += name + " = NULL, ";
            }
            setStmt += "sql_deleted = 0";

            // Create the where statement
            StringBuilder uWhereStmtBuilder = new StringBuilder("WHERE " + key + " IN (");
            for (Object col : cols) {
                uWhereStmtBuilder.append(col).append(",");
            }
            if (uWhereStmtBuilder.toString().endsWith(",")) {
                uWhereStmtBuilder.deleteCharAt(uWhereStmtBuilder.length() - 1);
            }
            uWhereStmtBuilder.append(");");
            uWhereStmt = uWhereStmtBuilder.toString();
        } catch (Exception error) {
            String msg = error.getMessage() != null ? error.getMessage() : "";
            throw new Exception(msg);
        }

        return new UpdateResults(setStmt, uWhereStmt);
    }

    public static void executeUpdateForDelete(
        Database mDB,
        String tableName,
        String whereStmt,
        String setStmt,
        List<String> colNames,
        ArrayList<Object> values
    ) throws Exception {
        try {
            long lastId = -1;

            // Update sql_deleted for this references
            String stmt = "UPDATE " + tableName + " SET " + setStmt + " " + whereStmt;
            ArrayList<Object> selValues = getSelectedValues(values, whereStmt, colNames);

            JSObject retObj = mDB.prepareSQL(stmt, selValues, false, "no");
            lastId = retObj.getLong("lastId");
            if (lastId == -1) {
                String msg = "UPDATE sql_deleted failed for table: " + tableName;
                throw new Exception(msg);
            }
        } catch (Exception error) {
            String msg = error.getMessage() != null ? error.getMessage() : "";
            throw new Exception(msg);
        }
    }

    public static ArrayList<Object> getSelectedValues(ArrayList<Object> values, String whereStmt, List<String> colNames) {
        ArrayList<Object> selValues = new ArrayList<>(); // Initialize the selected values ArrayList

        if (values.size() > 0) {
            String[] arrVal = whereStmt.split("\\?");
            if (arrVal[arrVal.length - 1].equals(";")) {
                arrVal[arrVal.length - 1] = "";
            }

            for (int jdx = 0; jdx < arrVal.length; jdx++) {
                for (String updVal : colNames) {
                    List<Integer> indices = UtilsSQLStatement.indicesOf(arrVal[jdx], updVal, 0);
                    if (!indices.isEmpty()) {
                        selValues.add(values.get(jdx));
                    }
                }
            }
        }

        return selValues;
    }
}

File: ios/Plugin/Utils/UtilsDelete.swift
//
//  UtilsDelete.swift
//  CapacitorCommunitySqlite
//
//  Created by  Quéau Jean Pierre on 30/07/2023.
//

import Foundation
enum UtilsDeleteError: Error {
    case findReferencesAndUpdate(message: String)
    case getReferences(message: String)
    case getRefs(message: String)
    case getUpdDelReturnedValues(message: String)
    case getFirstPK(message: String)
    case extractColumnNames(message: String)
    case executeUpdateForDelete(message: String)
    case extractForeignKeyInfo(message: String)
    case searchForRelatedItems(message: String)
    case upDateWhereForRestrict(message: String)
    case upDateWhereForCascade(message: String)
}

// swiftlint:disable file_length
// swiftlint:disable type_body_length
class UtilsDelete {

    // MARK: - findReferencesAndUpdate

    // swiftlint:disable function_body_length
    // swiftlint:disable cyclomatic_complexity
    class func findReferencesAndUpdate(mDB: Database, tableName: String,
                                       whereStmt: String,
                                       initColNames: [String],
                                       values: [Any]) throws -> Bool {
        do {
            let retBool: Bool = true
            let result = try getReferences(mDB: mDB,
                                           tableName: tableName)
            let references = result.retRefs
            let tableNameWithRefs = result.tableWithRefs
            if references.count <= 0 {
                return retBool
            }
            if tableName == tableNameWithRefs {
                return retBool
            }
            // Loop through references
            for ref in references {

                // Extract the FOREIGN KEY constraint info
                // from the ref statement
                let foreignKeyInfo = try UtilsSQLStatement
                    .extractForeignKeyInfo(from: ref)
                // get the tableName of the references
                guard let refTable = foreignKeyInfo["tableName"]
                        as? String else {
                    let msg = "findReferencesAndUpdate: no foreignKeyInfo " +
                        "tableName"
                    throw UtilsDeleteError
                    .findReferencesAndUpdate(message: msg)
                }
                if refTable.isEmpty || refTable != tableName {
                    continue
                }
                // get the with ref columnName
                guard let withRefsNames = foreignKeyInfo["forKeys"]
                        as? [String] else {

                    let msg = "findReferencesAndUpdate: no foreignKeyInfo " +
                        "forKeys"
                    throw UtilsDeleteError
                    .findReferencesAndUpdate(message: msg)
                }
                guard let colNames = foreignKeyInfo["refKeys"]
                        as? [String] else {
                    let msg = "findReferencesAndUpdate: no foreignKeyInfo " +
                        "refKeys"
                    throw UtilsDeleteError
                    .findReferencesAndUpdate(message: msg)
                }
                if colNames.count != withRefsNames.count {
                    let msg = "findReferencesAndUpdate: no foreignKeyInfo " +
                        "colNames"
                    throw UtilsDeleteError
                    .findReferencesAndUpdate(message: msg)

                }
                guard let action = foreignKeyInfo["action"]
                        as? String else {
                    let msg = "findReferencesAndUpdate: no action"
                    throw UtilsDeleteError
                    .findReferencesAndUpdate(message: msg)
                }
                if action == "NO_ACTION" {
                    continue
                }

                let updTableName: String = tableNameWithRefs
                let updColNames: [String] = withRefsNames
                var results: (setStmt: String, uWhereStmt: String)
                results.uWhereStmt = ""
                results.setStmt = ""
                if !checkValuesMatch(withRefsNames,
                                     against: initColNames) {
                    // Search for related items in tableName
                    let result: (String, [[String: Any]]) = try UtilsDelete
                        .searchForRelatedItems(mDB: mDB,
                                               updTableName: updTableName,
                                               tableName: tableName,
                                               whStmt: whereStmt,
                                               withRefsNames: withRefsNames,
                                               colNames: colNames,
                                               values: values)
                    let key: String = result.0
                    let relatedItems: [Any] = result.1
                    if relatedItems.count == 0 && key.count <= 0 {
                        continue
                    }

                    // case no match
                    if updTableName != tableName {
                        switch action {
                        case "CASCADE":
                            // updTableName
                            // update all related element
                            // set sql_deleted = 1 and last_modified
                            // tableName
                            // update all by sending return true
                            results = try upDateWhereForCascade(
                                results: result)

                        case "RESTRICT":
                            // find for elements related in updTableName
                            // if some elements
                            // send a message
                            // do not update tableName
                            // return false
                            // If no related elements in updTableName
                            // return true to update tableName
                            results = try upDateWhereForRestrict(
                                results: result)

                        default:
                            // updTableName
                            // update the result_id result_slug to Null
                            // update the last_modified
                            // keep sql_deleted to 0
                            // return true to update tableName
                            results = try upDateWhereForDefault(
                                withRefsNames: withRefsNames,
                                results: result)
                        }
                    }

                } else {
                    let msg = "Not implemented. Please transfer your " +
                        "example to the maintener"
                    throw UtilsDeleteError
                    .findReferencesAndUpdate(message: msg)
                }
                if results.setStmt.count > 0 &&
                    results.uWhereStmt.count > 0 {

                    try executeUpdateForDelete(
                        mDB: mDB,
                        tableName: updTableName,
                        whereStmt: results.uWhereStmt,
                        setStmt: results.setStmt,
                        colNames: updColNames,
                        values: values)
                }
            }
            return retBool
        } catch UtilsDeleteError.upDateWhereForRestrict(let message) {
            throw UtilsDeleteError
            .findReferencesAndUpdate(message: message)
        } catch UtilsSQLStatementError
                    .extractForeignKeyInfo(let message) {
            throw UtilsDeleteError
            .findReferencesAndUpdate(message: message)
        } catch UtilsDeleteError.executeUpdateForDelete(let message) {
            throw UtilsDeleteError
            .findReferencesAndUpdate(message: message)
        } catch UtilsSQLCipherError.prepareSQL(let message) {
            throw UtilsDeleteError
            .findReferencesAndUpdate(message: message)
        } catch UtilsSQLCipherError.querySQL(let message) {
            throw UtilsDeleteError
            .findReferencesAndUpdate(message: message)
        }
    }
    // swiftlint:enable cyclomatic_complexity
    // swiftlint:enable function_body_length

    // MARK: - upDateWhereForDefault

    class func upDateWhereForDefault(withRefsNames: [String],
                                     results: (String, [[String: Any]]))
    throws -> ((setStmt: String, uWhereStmt: String)) {

        var setStmt = ""
        var uWhereStmt = ""
        let key: String = results.0
        let relatedItems = results.1

        var cols: [Any] = []
        for relItem in relatedItems {
            if let mVal = relItem[key] {
                cols.append(mVal)
            }
        }
        // create the set statement
        for name in withRefsNames {
            setStmt += "\(name) = NULL, "
        }
        setStmt += "sql_deleted = 0"

        // create the where statement
        uWhereStmt = "WHERE \(key) IN ("
        for col in cols {
            uWhereStmt += "\(col),"
        }
        if uWhereStmt.hasSuffix(",") {
            uWhereStmt = String(uWhereStmt.dropLast())
        }

        uWhereStmt += ");"

        return(setStmt, uWhereStmt)
    }

    // MARK: - upDateWhereForRestrict

    class func upDateWhereForRestrict(results: (String, [[String: Any]]))
    throws -> ((setStmt: String, uWhereStmt: String)) {

        // Search for related items in tableName
        let setStmt = ""
        let uWhereStmt = ""
        let relatedItems = results.1

        if !relatedItems.isEmpty {
            let msg = "Restrict mode related items exist" +
                " please delete them first"
            throw UtilsDeleteError
            .upDateWhereForRestrict(message: msg)
        }
        return(setStmt, uWhereStmt)
    }

    // MARK: - upDateWhereForCascade

    class func upDateWhereForCascade(results: (String, [[String: Any]]))
    throws -> ((setStmt: String, uWhereStmt: String)) {

        // Search for related items in tableName
        var setStmt = ""
        var uWhereStmt = ""
        let key: String = results.0
        let relatedItems = results.1
        var cols: [Any] = []
        for relItem in relatedItems {
            if let mVal = relItem[key] {
                cols.append(mVal)
            }
        }
        setStmt += "sql_deleted = 1"
        // create the where statement
        uWhereStmt = "WHERE \(key) IN ("
        for col in cols {
            uWhereStmt += "\(col),"
        }
        if uWhereStmt.hasSuffix(",") {
            uWhereStmt = String(uWhereStmt.dropLast())
        }

        uWhereStmt += ");"
        return (setStmt, uWhereStmt)
    }

    // MARK: - searchForRelatedItems

    // swiftlint:disable function_parameter_count
    class func searchForRelatedItems(mDB: Database,
                                     updTableName: String,
                                     tableName: String, whStmt: String,
                                     withRefsNames: [String],
                                     colNames: [String], values: [Any])
    throws -> (String, [[String: Any]]) {
        var relatedItems: [[String: Any]] = []
        var key: String = ""
        let t1Names = withRefsNames.map({ "t1.\($0)" })
        let t2Names = colNames.map({ "t2.\($0)"})

        do {
            var whereClause = try UtilsSQLStatement
                .addPrefixToWhereClause(whStmt, from: colNames,
                                        destination: withRefsNames,
                                        prefix: "t2.")
            if whereClause.hasSuffix(";") {
                whereClause = String(whereClause.dropLast())
            }
            let resultString = zip(t1Names, t2Names)
                .map({ "\($0) = \($1)"})
                .joined(separator: " AND ")
            let sql = "SELECT t1.rowid FROM \(updTableName) t1 " +
                "JOIN \(tableName) t2 ON \(resultString) " +
                "WHERE \(whereClause) AND t1.sql_deleted = 0;"
            var vals = try UtilsSQLCipher.querySQL(mDB: mDB, sql: sql,
                                                   values: values)
            if vals.count > 1 {
                if let mVals = vals[0]["ios_columns"] as? [String] {
                    key = mVals[0]
                    let keyToRemove = "ios_columns"

                    // Remove dictionaries where the keys contain "ios_columns"
                    vals.removeAll { (dict: [String: Any]) in
                        return dict.keys.contains(keyToRemove)
                    }
                    // Append the remaining dictionaries to relatedItems
                    for val in vals {
                        relatedItems.append(val)
                    }
                }
            }
            return (key, relatedItems)
        } catch UtilsSQLStatementError.addPrefixToWhereClause(let message) {
            throw UtilsDeleteError
            .searchForRelatedItems(message: message)
        } catch UtilsSQLCipherError.querySQL(let message) {
            throw UtilsDeleteError
            .searchForRelatedItems(message: message)
        }
    }
    // swiftlint:enable function_parameter_count

    // MARK: - executeUpdateForDelete

    // swiftlint:disable function_parameter_count
    class func executeUpdateForDelete(mDB: Database, tableName: String,
                                      whereStmt: String, setStmt: String,
                                      colNames: [String], values: [Any])
    throws {
        var lastId: Int64 = -1
        // update sql_deleted for this references
        let stmt = "UPDATE \(tableName) SET \(setStmt) \(whereStmt)"
        var selValues: [Any] = []
        if !values.isEmpty {
            var arrVal: [String] =  whereStmt
                .components(separatedBy: "?")
            if arrVal[arrVal.count - 1] == ";" {
                arrVal.removeLast()
            }
            for (jdx, val) in arrVal.enumerated() {
                for updVal in colNames {
                    let indices: [Int] = val.indicesOf(string: updVal)
                    if indices.count > 0 {
                        selValues.append(values[jdx])
                    }
                }
            }
        }

        let resp = try UtilsSQLCipher.prepareSQL(mDB: mDB, sql: stmt,
                                                 values: selValues,
                                                 fromJson: false, returnMode: "no")
        lastId = resp.0
        if lastId == -1 {
            let msg = "UPDATE sql_deleted failed for " +
                "table: \(tableName) "
            throw UtilsDeleteError.executeUpdateForDelete(message: msg)
        }

    }

    // MARK: - getCurrentTimeAsInteger

    class func getCurrentTimeAsInteger() -> Int {
        let currentTime = Date().timeIntervalSince1970
        return Int(currentTime)
    }

    // MARK: - checkValuesMatch

    class func checkValuesMatch(_ array1: [String],
                                against array2: [String]) -> Bool {
        for value in array1 {
            if !array2.contains(value) {
                return false
            }
        }
        return true
    }

    // MARK: - getReferences

    class func getReferences(mDB: Database, tableName: String)
    throws -> (tableWithRefs: String, retRefs: [String]) {
        // find the REFERENCES
        var sqlStmt = "SELECT sql FROM sqlite_master "
        sqlStmt += "WHERE sql LIKE('%FOREIGN KEY%') AND "
        sqlStmt += "sql LIKE('%REFERENCES%') AND "
        sqlStmt += "sql LIKE('%\(tableName)%') AND "
        sqlStmt += "sql LIKE('%ON DELETE%');"
        do {
            var references: [[String: Any]] = try UtilsSQLCipher
                .querySQL(mDB: mDB, sql: sqlStmt, values: [])
            var retRefs = [String]()
            var tableWithRefs: String = ""
            if references.count > 1 {
                references.removeFirst()
                if let refValue = references[0]["sql"] as? String {
                    let result = try getRefs(sqlStatement: refValue)
                    retRefs = result.foreignKeys
                    tableWithRefs = result.tableName
                }
            }
            return (tableWithRefs, retRefs)
        } catch UtilsDeleteError.getRefs(let message) {
            throw UtilsDeleteError
            .getReferences(message: message)
        } catch UtilsSQLCipherError.querySQL(let message) {
            throw UtilsDeleteError
            .getReferences(message: message)
        }
    }

    // MARK: - getRefs

    class func getRefs(sqlStatement: String)
    throws -> (tableName: String, foreignKeys: [String]) {
        var tableName = ""
        var foreignKeys = [String]()
        let statement = UtilsSQLStatement
            .flattenMultilineString(sqlStatement)

        do {
            // Regular expression pattern to match the table name
            let tableNamePattern = #"CREATE\s+TABLE\s+(\w+)\s+\("#
            let tableNameRegex = try NSRegularExpression(
                pattern: tableNamePattern, options: [])
            if let tableNameMatch = tableNameRegex
                .firstMatch(in: statement, options: [],
                            range: NSRange(location: 0,
                                           length: statement.utf16.count)) {
                if let tableNameRange = Range(tableNameMatch.range(at: 1), in: statement) {
                    tableName = String(statement[tableNameRange])
                } else {
                    let msg = "getRefs: Error creating tableNameRange "
                    throw UtilsDeleteError.getRefs(message: msg)
                }
            } else {
                let msg = "getRefs: Error creating tableNameMatch "
                throw UtilsDeleteError.getRefs(message: msg)
            }

            // Regular expression pattern to match the FOREIGN KEY
            // constraints
            // swiftlint:disable line_length
            let foreignKeyPattern = #"FOREIGN\s+KEY\s+\([^)]+\)\s+REFERENCES\s+(\w+)\s*\([^)]+\)\s+ON\s+DELETE\s+(CASCADE|RESTRICT|SET\s+DEFAULT|SET\s+NULL|NO\s+ACTION)"#
            // swiftlint:enable line_length

            let foreignKeyRegex = try NSRegularExpression(
                pattern: foreignKeyPattern, options: [])
            let foreignKeyMatches = foreignKeyRegex
                .matches(in: statement, options: [],
                         range: NSRange(location: 0,
                                        length: statement.utf16.count))
            for foreignKeyMatch in foreignKeyMatches {
                if let foreignKeyRange = Range(
                    foreignKeyMatch.range(at: 0), in: statement) {
                    let foreignKey = String(statement[foreignKeyRange])
                    foreignKeys.append(foreignKey)
                } else {
                    let msg = "getRefs: Error creating foreignKeyRange "
                    throw UtilsDeleteError.getRefs(message: msg)
                }
            }
        } catch {
            let msg = "getRefs: Error creating regular expression: " +
                "\(error)"
            throw UtilsDeleteError.getRefs(message: msg)
        }
        return (tableName, foreignKeys)
    }

    // MARK: - getUpdDelReturnedValues

    class func getUpdDelReturnedValues(mDB: Database,
                                       sqlStmt: String,
                                       names: String )
    throws -> [[String: Any]] {
        var result: [[String: Any]] = []
        let tableName = UtilsSQLStatement
            .extractTableName(from: sqlStmt)
        let whereClause = UtilsSQLStatement
            .extractWhereClause(from: sqlStmt)
        if let tblName = tableName {
            if var wClause = whereClause {
                if wClause.suffix(1) == ";" {
                    wClause = String(wClause.dropLast())
                }
                do {
                    var query: String = "SELECT \(names) FROM " +
                        "\(tblName) WHERE "
                    query += "\(wClause);"
                    result = try UtilsSQLCipher
                        .querySQL(mDB: mDB, sql: query, values: [])
                } catch UtilsSQLCipherError.querySQL(let message) {
                    throw UtilsDeleteError
                    .getUpdDelReturnedValues(message: message)
                }
            }
        }
        return result
    }
}
// swiftlint:enable type_body_length
// swiftlint:enable file_length
